# LOGISTIC REGRESSION

**In logistic regression we can train our data to get values(result) between 0-1**

In [41]:
import numpy as np
import pandas as pd
dataset=pd.read_csv('patterned_user_data.csv')
dataset

Age  EstimatedSalary  Purchased
0    25            64758          0
1    38           112409          1
2    47           111211          0
3    33           105697          1
4    29            56065          1
..  ...              ...        ...
95   40           115766          1
96   28            34707          0
97   22            40976          0
98   40            84262          1
99   47            63776          1

[100 rows x 3 columns]

In [42]:
X=dataset.iloc[:, :-1].values

In [43]:
X


array([[    25,  64758],
       [    38, 112409],
       [    47, 111211],
       [    33, 105697],
       [    29,  56065],
       [    26,  51606],
       [    47,  51534],
       [    39, 134663],
       [    25,  59397],
       [    44, 131387],
       [    37,  41016],
       [    41, 129789],
       [    29,  74591],
       [    29,  42247],
       [    42,  64300],
       [    39, 114065],
       [    22,  28268],
       [    26,  31185],
       [    42, 103704],
       [    21,  58099],
       [    40,  48571],
       [    39, 145512],
       [    20,  57044],
       [    42,  91214],
       [    30,  80228],
       [    48,  88984],
       [    24,  59774],
       [    20,  21568],
       [    50, 102592],
       [    46, 107563],
       [    39,  42695],
       [    19,  67190],
       [    30,  24258],
       [    44, 127538],
       [    40,  79504],
       [    47,  73159],
       [    30,  32986],
       [    43, 101858],
       [    35,  52666],
       [    45,  78660],


In [44]:
y=dataset.iloc[:, -1].values

In [45]:
y

array([0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1,
       0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 0,
       1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1], dtype=int64)

In [46]:
from sklearn.model_selection import train_test_split

In [47]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.25,random_state=0)

In [48]:
len(X_train)

75

In [49]:
len(X_test)

25

In [50]:
len(y_train)

75

In [51]:
len(y_test)

25

**The accuracy of the data sets can be improved if we feature scale the data further**

### Splitting Data

In [52]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()

In [53]:
X_train=sc.fit_transform(X_train)
X_test=sc.transform(X_test)

In [54]:
X_train

array([[ 1.36059277,  1.22047442],
       [ 1.25260922, -0.7804414 ],
       [ 1.25260922, -0.41521735],
       [ 1.25260922,  2.1340417 ],
       [ 0.28075724,  0.74373204],
       [ 0.17277369,  0.45640376],
       [-1.12302895, -1.08578922],
       [-1.0150454 , -0.29143737],
       [ 1.14462566,  0.49447155],
       [ 1.36059277, -0.07374093],
       [-1.66294672, -1.67443778],
       [ 0.71269145,  0.77598225],
       [ 1.57655988,  2.09946445],
       [-0.90706185, -1.74490496],
       [ 0.38874079, -0.11535893],
       [ 0.28075724,  1.03568453],
       [-1.12302895, -1.54489989],
       [-0.79907829,  0.17337153],
       [ 1.46857632, -0.32338925],
       [-0.69109474, -0.64526494],
       [ 0.38874079,  1.08508913],
       [-1.0150454 , -1.38752717],
       [ 1.03664211, -1.01830541],
       [-0.04319342, -0.74666966],
       [-1.0150454 , -0.77829337],
       [-1.44697961, -0.46996218],
       [-1.33899606, -0.93760528],
       [-1.12302895, -0.38592067],
       [ 0.49672434,

**Now we can observe both the colums of X_train are in similar data type(digits)**

In [55]:
X_test

array([[-1.23101251, -0.53461178],
       [-0.36714408,  2.14391665],
       [ 1.25260922,  0.99994377],
       [ 0.28075724, -0.39248408],
       [-1.66294672, -0.08573407],
       [ 1.36059277, -0.58613457],
       [-1.44697961, -1.47455205],
       [ 0.92865856,  0.88069851],
       [ 0.6047079 , -0.79741677],
       [ 0.49672434,  1.13583623],
       [-0.58311119,  0.13846611],
       [-1.66294672, -1.09966189],
       [-0.25916053,  1.12640879],
       [-0.69109474, -1.05750688],
       [ 0.38874079,  1.69960349],
       [ 0.38874079, -1.04414139],
       [-1.66294672, -0.61605776],
       [-0.58311119,  0.07560653],
       [ 0.92865856,  1.48703844],
       [-1.12302895, -0.54585909],
       [ 1.57655988,  1.9927195 ],
       [ 0.38874079,  1.58978567],
       [-0.25916053,  0.8354408 ],
       [ 0.92865856,  1.27277288],
       [ 1.14462566, -0.44845209]])

### Model

In [56]:
from sklearn.linear_model import LogisticRegression

In [57]:
classifier = LogisticRegression(random_state=0)

In [58]:
classifier.fit(X_train,y_train)

LogisticRegression(random_state=0)

### Prediction

In [60]:
print(classifier.predict(sc.transform([[39,134663]]))[0])

1


In [64]:
print(classifier.predict(sc.transform([[44,120356]]))[0])


1


In [67]:
y_pred=classifier.predict(X_test)

In [68]:
y_pred

array([0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1,
       1, 1, 0], dtype=int64)

In [71]:
print(np.concatenate((y_pred.reshape(len(y_pred),1),y_test.reshape(len(y_test),1)),1))

[[0 0]
 [1 1]
 [1 0]
 [0 0]
 [0 0]
 [0 1]
 [0 0]
 [1 0]
 [0 0]
 [1 1]
 [0 1]
 [0 0]
 [1 1]
 [0 0]
 [1 0]
 [0 0]
 [0 0]
 [0 1]
 [1 1]
 [0 1]
 [1 1]
 [1 0]
 [1 1]
 [1 1]
 [0 0]]


**actual value v/s predicted value as we can see our model is not able to predict properly as the data we have provided is not very accurate or linear**

In [72]:
from sklearn.metrics import confusion_matrix,accuracy_score

In [73]:
cm=confusion_matrix(y_test,y_pred)

In [74]:
cm

array([[10,  4],
       [ 4,  7]], dtype=int64)

In [75]:
accuracy_score(y_test,y_pred)

0.68

### Thus we get an accuracy of 0.68 , this shows how accurate our model is in predicting